frontends/p4/typeChecking/typeUnification.cpp

```diff
@@ -296,32 +295,73 @@ bool TypeUnification::unify(const EqualityConstraint* constraint) {
         if (!src->is<IR::Type_BaseList>())
             return constraint->reportError(constraints->getCurrentSubstitution());
         auto ts = src->to<IR::Type_BaseList>();
-        if (td->components.size() != ts->components.size())
-            return constraint->reportError(
-                constraints->getCurrentSubstitution(),
-                "Tuples with different sizes %1% vs %2%",
-                td->components.size(), ts->components.size());
-        for (size_t i=0; i < td->components.size(); i++) {
-            auto si = ts->components.at(i);
-            auto di = td->components.at(i);
-            constraints->add(constraint->create(di, si));
+        bool defaultInitializer = false;
+        if (auto lst = src->to<IR::Type_List>()) {
+            defaultInitializer = lst->defaultInitializer;
+        }
+        if (td->components.size() != ts->components.size() && !defaultInitializer) {
+            return constraint->reportError(constraints->getCurrentSubstitution(),
+            "Tuples with different sizes %1% vs %2%",td->components.size(), ts->components.size());
+        }
+        if (!defaultInitializer) {
+            for (size_t i=0; i < td->components.size(); i++) {
+                auto si = ts->components.at(i);
+                auto di = td->components.at(i);
+                constraints->add(constraint->create(di, si));
+            }
+        } else {
+            for (size_t i=0; i < td->components.size(); i++) {
+                auto di = td->components.at(i);
+                if (di->is<IR::Type_Varbits>() || di->is<IR::Type_Stack>()) {
+                    return constraint->reportError(constraints->getCurrentSubstitution(),
+                                                "Structs, headers and tuples "
```

Comment:
frankly this is not really the job of the type unification. that error should be given somewhere else.
type unification should only say whether the two types can be made equal. moreover, if we decide to have default initializers for these types this code should probably not change.